### ISSUE TITLE
Exporting reports causes API to stop

### ISSUE THREAD

**Issue by sglangevin** [medic/medic-webapp#4596]
When attempting to export all reports (no filters selected) on beta-old.dev, the file starts downloading but stops and never completes. It also causes API to stop with no errors in the logs. This is working properly on standard.dev which is also on the same version but has fewer reports.

**Steps to reproduce**:
 - Log into beta-old.dev as admin
 - Go to the reports page
 - Click Export

**What should happen**:
 - CSV should download that contains all reports, all fields

**What actually happens**:
 - CSV starts downloading but doesn't finish - it stops at 251 KB each time
 - API stops with no errors in the logs

https://beta-old.dev.medicmobile.org/medic/_design/medic/_rewrite/#/reports/

I couldn't find anything in the logs, but here are the results from `top` - lots of CPU usage.
```
Mem: 3325280K used, 337604K free, 4376K shrd, 72540K buff, 2421808K cached
CPU: 97.7% usr  2.2% sys  0.0% nic  0.0% idle  0.0% io  0.0% irq  0.0% sirq
Load average: 1.32 0.60 0.37 3/177 12014
```

| PID | PPID | USER | STAT | VSZ | %VSZ | CPU | %CPU | COMMAND |
| ---- | ------ | ------- | ------ | ---- | -------- | ----- | -------- | -------------- |
| 11845 | 11266 | vm | R | 11292 | 0.3 | 0 | 0.2 | top |
| 11265 | 11052 | vm | S | 21172 | 0.5 | 0 | 0.0 | sshd: vm@pts/0 |
| 8105 | 1 | vm | S | 12860 | 0.3 | 0 | 0.0 | {screen} SCREEN -R |
| 8106 | 8105 | vm | S | 11292 | 0.3 | 0 | 0.0 | /bin/sh |
| 11266 | 11265 | vm | S | 11292 | 0.3 | 0 | 0.0 | -sh |

**Environment**:

- Instance: beta-old.dev
- Browser: Chrome
- Client platform: MacOS
- App: webapp
- Version: 2.15.0-beta.15

**Other**:
API is definitely stopping because restarting services brings it back up and the instance starts working again. Probably worth investigating whether there is something specific wrong with the beta-old instance or if this is happening elsewhere as well. I did make sure to delete API and Sentinel before upgrading to the latest beta but maybe there is something else going on. Since this is working fine on standard.dev, it could be a server issue. It was also weird that the export from standard.dev was exactly 250 KB and the export from beta-old.dev stops at exactly 251 KB each time.

**Comment by SCdF** [comment]
Yep, this happens due to bad data, with a person with itself as the parent. Interestingly enough this person is marked as `"notes": "Purposefully has no parent",` so I wonder who did that, how they did that and how this then caused our software to make themselves the parent.

@garethbowen did we come to a conclusion on how we want to deal with this? We can put a recursion depth check in and throw our toys if something is clearly broken, or we could ignore it, or...?

**Comment by SCdF** [comment]
Specifically it's this report: http://localhost:15984/_utils/#database/medic/5a5298b0a308146efa4ab51d74ad6a30 who points to this person: http://localhost:15984/_utils/#database/medic/8e0ff9fd-ff6a-442e-ad2d-ed7c915838dd

**Comment by garethbowen** [comment]
@dianabarsan fixed the similar issue in #4487 in 3.0.0 by patching the `docs_by_id_lineage` view. Is that view used? If so, we could look at backporting the fix. Otherwise, I'm not too worried about this as I suspect it won't happen in the wild, and if it does we can fix the data to resolve it without release.

**Comment by SCdF** [comment]
@garethbowen I added a guard, see the PR. I think this is good enough for 2.15

**Comment by garethbowen** [comment]
Agreed - good enough for 2.15.

I'll leave it to you to merge to `2.15.x`, forward port to `2.16.x`, and maybe forward port the error handling change to `master`?

**Comment by SCdF** [comment]
@garethbowen merged to 2.15 and 2.16. In the interest of working on 2.16 I raised https://github.com/medic/medic-webapp/issues/4604 in 3.0 and I'll mark as it a requirement for release, so we don't forget / deprioritise it

**Comment by sglangevin** [comment]
LGTM and it even helped me find the report with the circular reference so that I could delete it. Moving to ready.
